snuba/migrations/snuba_migrations/metrics/0020_polymorphic_buckets_table.py

```diff
@@ -0,0 +1,74 @@
+from typing import Sequence
+
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.utils.schemas import Array, Column, DateTime, Float, Nested, String, UInt
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    """
+    Create a polymorphic bucket table for metrics_sets,
+    metrics_distributions, and metrics_counters
+    """
+
+    blocking = False
+    local_table_name = "metrics_buckets_consolidated"
+    dist_table_name = "metrics_buckets_consolidated_dist"
+
+    column_list: Sequence[Column[Modifiers]] = [
+        Column("org_id", UInt(64)),
+        Column("project_id", UInt(64)),
+        Column("metric_id", UInt(64)),
+        Column("timestamp", DateTime()),
+        Column("tags", Nested([Column("key", UInt(64)), Column("value", UInt(64))]),),
+        Column("metric_type", String(Modifiers(low_cardinality=True))),
+        Column("set_values", Array(UInt(64))),
+        Column("count_value", Float(64)),
+        Column("distribution_values", Array(Float(64))),
+        Column("materialization_version", UInt(8)),
+        Column("retention_days", UInt(16)),
+        Column("partition", UInt(16)),
+        Column("offset", UInt(64)),
+    ]
+
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.METRICS,
+                table_name=self.local_table_name,
+                columns=self.column_list,
+                engine=table_engines.MergeTree(
+                    storage_set=StorageSetKey.METRICS,
+                    order_by="(project_id, metric_id, tags.key, tags.value, timestamp)",
+                    partition_by="toStartOfDay(timestamp)",
+                    ttl="timestamp + toIntervalDay(7)",
+                ),
+            )
+        ]
+
+    def backwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.DropTable(
+                storage_set=StorageSetKey.METRICS, table_name=self.local_table_name
+            )
+        ]
+
+    def forwards_dist(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.METRICS,
+                table_name=self.dist_table_name,
+                columns=self.column_list,
+                engine=table_engines.Distributed(
+                    local_table_name=self.local_table_name, sharding_key=None,
```

Comment:
Should we add a sharding key for the table? Maybe by the `org_id` or `project_id`? That way the materialized view does not need to go across shards when we increase the number of shards.

Reply:
I was sort of copying this forward from another migration: https://github.com/getsentry/snuba/blob/b3d58e642dd1d1dee53a1cbe022e3dde159e38ac/snuba/migrations/snuba_migrations/metrics/0001_metrics_buckets.py#L25

It's not clear to me what the distributed table really does for us here, but maybe @fpacifici could add a comment since he wrote that migration

Reply:
> Should we add a sharding key for the table? Maybe by the org_id or project_id? That way the materialized view does not need to go across shards when we increase the number of shards.

The materialized view only operates locally on local tables, it would never go across shards.

> It's not distributing application reads which are all going through the aggregate tables, and I assume writes on any local instance causes all the replicas to make the same copies

If you write directly on a storage node the distributed table is not needed, but, since we are likely to reintroduce the query node as we will nto have time to fix the deployment problems before production without the query node, we will need a distributed table for the query node to write onto.